ProcessLocksScreen.c

```diff
@@ -0,0 +1,247 @@
+/*
+htop - ProcessLocksScreen.c
+(C) 2005-2006 Hisham H. Muhammad
+Released under the GNU GPLv2, see the COPYING file
+in the source distribution for its full text.
+*/
+
+
+#include "config.h" // IWYU pragma: keep
+#include "CRT.h"
+#include "FunctionBar.h"
+#include "IncSet.h"
+#include "ProcessList.h"
+#include "ProcessLocksScreen.h"
+#include "XUtils.h"
+
+#include <dirent.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+
+typedef struct FileLocks_Data_ {
+   char* data[4];
+   int id;
+   unsigned int dev[2];
+   uint64_t inode;
+   uint64_t start;
+   uint64_t end;
+} FileLocks_Data;
+
+typedef struct FileLocks_ProcessData_ {
+   int error;
+   struct FileLocks_LockData_* locks;
+} FileLocks_ProcessData;
+
+typedef struct FileLocks_LockData_ {
+   FileLocks_Data data;
+   struct FileLocks_LockData_* next;
+} FileLocks_LockData;
+
+ProcessLocksScreen* ProcessLocksScreen_new(const Process* process) {
+   ProcessLocksScreen* this = xMalloc(sizeof(ProcessLocksScreen));
+   Object_setClass(this, Class(ProcessLocksScreen));
+   if (Process_isThread(process))
+      this->pid = process->tgid;
+   else
+      this->pid = process->pid;
+   return (ProcessLocksScreen*) InfoScreen_init(&this->super, process, NULL, LINES-3, "        ID  TYPE       EXCLUSION  READ/WRITE DEVICE:INODE                              START                  END  FILENAME");
+}
+
+void ProcessLocksScreen_delete(Object* this) {
+   free(InfoScreen_done((InfoScreen*)this));
+}
+
+static void ProcessLocksScreen_draw(InfoScreen* this) {
+   InfoScreen_drawTitled(this, "Snapshot of file locks of process %d - %s", ((ProcessLocksScreen*)this)->pid, this->process->comm);
+}
+
+/*
+ * Return the absolute path of a file given its pid&inode number
+ *
+ * Based on implementation of lslocks from util-linux:
+ * https://sources.debian.org/src/util-linux/2.36-3/misc-utils/lslocks.c/#L162
+ */
+static char *ProcessLocksScreen_getInodeFilename(pid_t pid, ino_t inode) {
+   struct stat sb;
+   struct dirent *de;
+   DIR *dirp;
+   size_t len;
+   int fd;
+
+   char path[PATH_MAX];
+   char sym[PATH_MAX];
+   char* ret = NULL;
+
+   memset(path, 0, sizeof(path));
+   memset(sym, 0, sizeof(sym));
+
+   /*
+    * We know the pid so we don't have to
+    * iterate the *entire* filesystem searching
+    * for the damn file.
```

Comment:
damn - let's keep the language clean please (your mum might read this code someday).

Reply:
Unlikely, but removed anyway …